[ST-NNNN] Polling Confirmations (#2926)
apple/swift-evolution: proposals/testing/NNNN-polling-confirmations.md

  try await Task.sleep(for: .seconds(1))

  // Assert
  let data = try Data(contentsOf: logURL)
  let decoder = JSONLinesDecoder()
  // Implementation of `JSONLinesDecoder` left as an exercise to the reader.
  decoder.dateDecodingStrategy = .iso8601
  let messages = try decoder.decoding(Log.self, from: data)

  #expect(messages.count == 1)
}
```

In this case, in the failure case where 2 near-identical logs are incorrectly
recorded, then sleeping will still unnecessarily delay the test suite
execution. The test should fail fast instead of delaying any more than
absolutely necessary.

This proposal introduces polling to help test authors address these cases. In
this and other similar cases, polling makes these classes of tests practical or
even possible, as well as speeding up the execution of individual tests as well
as the entire test suite.

## Proposed solution

This proposal introduces new members of the `confirmation` family of functions:
`confirmation(_:until:within:pollingEvery:sourceLocation:_:)`. These
functions take in a closure to be repeatedly evaluated until the specific
condition passes, waiting at least some amount of time - specified by

> `pollingEvery`/`interval` and defaulting to 1 millisecond - before evaluating

Reviewer: I think it would be better to not explicitly specify what the default polling interval will be, because it may need to vary per-platform or be dynamic (which is related to the idea I suggested a while back which led to the "Curved polling rates" future direction discussed below).

In particular, I think 1 millisecond may be too aggressive in practice, especially when parallelization is enabled. By comparison, XCTest (in its [`XCTWaiter` implementation](https://github.com/swiftlang/swift-corelibs-xctest/blob/cc562f728ece742925b75216ee5a16748655fea8/Sources/XCTest/Public/Asynchronous/XCTWaiter.swift#L425)) uses a polling interval of 0.1 seconds, i.e. 100ms, and it never even has more than one test running at a time in the same process.

So while I think the implementation of this feature likely should poll less frequently than 1ms, when it comes to this _proposal_ my feedback is that a default value should simply be left undefined. The proposal could say something less committal instead, like,

> The specific default polling interval is platform- and environment-dependent, and may change dynamically in response to system load. But its default value will typically be 1 second or less.

///
/// - Throws: A ``PollingFailedError`` if the `body` does not return true within
///   the polling duration.
///
/// Use polling confirmations to check that an event while a test is running in
/// complex scenarios where other forms of confirmation are insufficient. For
/// example, waiting on some state to change that cannot be easily confirmed
/// through other forms of `confirmation`.
@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
public func confirmation(
  _ comment: Comment? = nil,
  until stopCondition: PollingStopCondition,
  within duration: Duration? = nil,
  pollingEvery interval: Duration? = nil,
  sourceLocation: SourceLocation = #_sourceLocation,
  _ body: nonisolated(nonsending) @escaping () async throws -> PollResult<Bool>
) async throws {
  let poller = Poller(
    stopCondition: stopCondition,
    duration: stopCondition.duration(with: duration),
    interval: stopCondition.interval(with: interval),
    comment: comment,
    sourceContext: SourceContext(
      backtrace: .current(),
      sourceLocation: sourceLocation
    )
  )
  try await poller.evaluate() {
    do {
      return try await body()
    } catch {
      return false
    }
  }

> }

Reviewer: The implementation of the function can be omitted from the proposal

///
/// - Returns: The last non-nil value returned by `body`.
///
/// Use polling confirmations to check that an event while a test is running in
/// complex scenarios where other forms of confirmation are insufficient. For
/// example, waiting on some state to change that cannot be easily confirmed
/// through other forms of `confirmation`.
@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
@discardableResult
public func confirmation<R>(
  _ comment: Comment? = nil,
  until stopCondition: PollingStopCondition,
  within duration: Duration? = nil,
  pollingEvery interval: Duration? = nil,
  sourceLocation: SourceLocation = #_sourceLocation,
  _ body: nonisolated(nonsending) @escaping () async throws -> sending PollResult<R>
) async throws -> R {
  let poller = Poller(
    stopCondition: stopCondition,
    duration: stopCondition.duration(with: duration),
    interval: stopCondition.interval(with: interval),
    comment: comment,
    sourceContext: SourceContext(
      backtrace: .current(),
      sourceLocation: sourceLocation
    )
  )
  return try await poller.evaluateOptional() {
    do {
      return try await body()
    } catch {
      return nil
    }
  }

> }

Reviewer: Function body can be omitted here as well

///     about 1 second before recording an issue.
///     `duration` must be greater than 0.
///   - interval: The minimum amount of time to wait between polling attempts.
///     If nil, this uses whatever value is specified under the last
///     ``PollingUntilFirstPassConfigurationTrait`` or
///     ``PollingUntilStopsPassingConfigurationTrait`` added to the test or
///     suite.
///     If no such trait has been added, then polling will wait at least
///     1 millisecond between polling attempts.
///     `interval` must be greater than 0.
///   - isolation: The actor to which `body` is isolated, if any.
///   - sourceLocation: The source location to whych any recorded issues should
///     be attributed.
///   - body: The function to invoke.
///
/// - Throws: A `PollingFailedError` if the `body` does not return true within
///   the polling duration.
///
/// Use polling confirmations to check that an event while a test is running in
/// complex scenarios where other forms of confirmation are insufficient. For
/// example, waiting on some state to change that cannot be easily confirmed
/// through other forms of `confirmation`.

> @available(_clockAPI, *)

Reviewer: `_clockAPI` is an availability definition specific to the Swift Testing codebase, and at build time it expands to actual OS version numbers. For the purpose of an API proposal I recommend showing the actual, expanded OS versions so people understand what they'll be in practice.

(This comment applies everywhere in the document which references `_clockAPI`.)

polling before the duration has elapsed. Additionally, if the early stop
condition isn't fulfilled before the duration elapses, then this also defines
how the confirmation should be handled.

```swift
/// A type defining when to stop polling early.
/// This also determines what happens if the duration elapses during polling.
public enum PollingStopCondition: Sendable {
  /// Evaluates the expression until the first time it returns true.
  /// If it does not pass once by the time the timeout is reached, then a
  /// failure will be reported.
  case firstPass

  /// Evaluates the expression until the first time it returns false.
  /// If the expression returns false, then a failure will be reported.
  /// If the expression only returns true before the timeout is reached, then
  /// no failure will be reported.
  /// If the expression does not finish evaluating before the timeout is
  /// reached, then a failure will be reported.
  case stopsPassing
}
```

### New Error Type

A new error type, `PollingFailedError` to be thrown when the polling
confirmation doesn't pass:

```swift
/// A type describing an error thrown when polling fails.

> public struct PollingFailedError: Error, Sendable, CustomIssueRepresentable {}

Reviewer: `CustomIssueRepresentable` is not a `public` protocol so it doesn't need to be shown here and I'd omit it to avoid confusion.

```

### New `Issue.Kind` case

A new issue kind will be added to report specifically when a test fails due to
a failed polling confirmation.

```swift
public struct Issue {

>   public enum Kind {

Reviewer: Small style suggestion: When proposing a new member of an existing type, such as a new case in an enum here, I often add a placeholder comment like `// ...` above the additions to reflect that there are existing members and you're just _adding_ things.

    ///
    /// This issue can occur when calling ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-455gr``
    /// or
    /// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-5tnlk``
    /// whenever the polling fails, as described in ``PollingStopCondition``.
    case pollingConfirmationFailed
  }
}
```

### New Traits

Two new traits will be added to change the default values for the
`duration` and `interval` arguments. Test authors will often want to poll for
the `firstPass` stop condition for longer than they poll for the
`stopsPassing` stop condition, which is why there are separate traits for
configuring defaults for these functions.

```swift
/// A trait to provide a default polling configuration to all usages of
/// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-455gr``
/// and
/// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-5tnlk``
/// within a test or suite for the ``PollingStopCondition.firstPass``
/// stop condition.
///
/// To add this trait to a test, use the
/// ``Trait/pollingUntilFirstPassDefaults`` function.
@available(_clockAPI, *)

> public struct PollingUntilFirstPassConfigurationTrait: TestTrait, SuiteTrait {

Reviewer: Assuming you do want to include the ability to influence this feature's behavior via a trait, I would suggest consolidating these two trait types into one if possible. For other traits we've tried to have the types be broader and use the `static` functions which initialize those types be the point where we differentiate their behavior. For example, the `.enabled(if:)` and `.disabled(if:)` functions both return a type of the same type: `ConditionTrait`.

But to accomplish this consolidation you'll likely need to store the `PollingStopCondition` in the trait instance. I think these changes would make for a simpler trait story for this feature: a single trait with a simpler name like `PollingConfigurationTrait`, with a property specifying its stopping condition.

  /// How long to continue polling for
  public var duration: Duration?
  /// The minimum amount of time to wait between polling attempts
  public var interval: Duration?

  public var isRecursive: Bool { true }
}

/// A trait to provide a default polling configuration to all usages of
/// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-455gr``
/// and
/// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-5tnlk``
/// within a test or suite for the ``PollingStopCondition.stopsPassing``
/// stop condition.
///
/// To add this trait to a test, use the ``Trait/pollingUntilStopsPassingDefaults``
/// function.
@available(_clockAPI, *)
public struct PollingUntilStopsPassingConfigurationTrait: TestTrait, SuiteTrait {
  /// How long to continue polling for
  public var duration: Duration?
  /// The minimum amount of time to wait between polling attempts

>   public var interval: Duration?

Reviewer: Since both of these properties are optional, can you document (and decide, if it hasn't been settled yet) what the behavior will be if either or both of these are `nil`?

>   public var isRecursive: Bool { true }

Reviewer: You can omit the `{ true }` _implementation_ and only show that it's a `get`-able property, e.g.

```suggestion
  public var isRecursive: Bool { get }
```

But I would recommend _documenting_ that this trait type is recursively applied to children. For other trait types we've done this in the DocC for the type itself.

}

@available(_clockAPI, *)
extension Trait where Self == PollingUntilFirstPassConfigurationTrait {

>   /// Specifies defaults for ``confirmPassesEventually`` in the test or suite.

Reviewer: This documentation still references the (older, I think?) API name `confirmPassesEventually`. (Similar comment for the other trait below, referencing `confirmPassesAlways`.)

  ///
  /// - Parameters:
  ///   - duration: The expected length of time to continue polling for.
  ///     This value may not correspond to the wall-clock time that polling
  ///     lasts for, especially on highly-loaded systems with a lot of tests
  ///     running.
  ///     if nil, polling will be attempted for approximately 1 second.
  ///     `duration` must be greater than 0.
  ///   - interval: The minimum amount of time to wait between polling
  ///     attempts.
  ///     If nil, polling will wait at least 1 millisecond between polling
  ///     attempts.
  ///     `interval` must be greater than 0.
  public static func pollingUntilStopsPassingDefaults(
    until duration: Duration? = nil,
    pollingEvery interval: Duration? = nil
  ) -> Self
}
```

Specifying `duration` or `interval` directly on either new `confirmation`
function will override any value provided by the relevant trait. Additionally,
when multiple of these configuration traits are specified, the innermost or
last trait will be applied.

### Default Polling Configuration

For all polling confirmations, the Testing library will default `duration` to

> 1 second, and `interval` to 1 millisecond.

Reviewer: I'm somewhat hesitant to formally document and "promise" particular values for the duration and polling interval. Moreso for the latter; those are effectively policy details and I could imagine in the future having those be platform- or environment-specific. Or we could get fancier and decide to make the polling interval a curve rather than a fixed value: for example it could poll very frequently and aggressively at the beginning, but over time gradually back off if it seems like the condition is taking a while to become satisfied.

These are just speculative ideas and I'm not suggesting you actually pursue them right now. (Maybe they'd be good to mention in Alternatives considered?) But for now, I'm only suggesting that if we formally specify and commit to those numbers in this proposal, we might lose the flexibility to refine the behavior later. So maybe it would be better to be a little more vague, and say that the default interval will be short, but the testing library may adjust it to behave well in each runtime environment.

### Platform Availability

Polling confirmations will not be available on platforms that do not support

> Swift Concurrency.

Reviewer: The availability will also be constrained to OS versions which have the `Clock`, `Duration`, and related types. You may want to mention the specific Apple OS versions in particular.

`duration` might elapse before the `body` has had a chance to return even once.

To prevent this, the Testing library will calculate how many times to poll the
`body`. This is done by dividing the `duration` by the `interval`. For example,
with the default 1 second duration and 1 millisecond interval, the Testing
library will poll 1000 times, waiting 1 millisecond between polling attempts.
This works and is immune to the issues posed by concurrent execution on
heavily-burdened systems.
This is also very easy for test authors to understand and predict, even if it is
not fully accurate - each poll attempt takes some amount of time, even for very
fast `body` closures. Which means that the real-time duration of a polling
confirmation will always be longer than the value specified in the `duration`
argument.

### Usage

These functions can be used with an async test function:

```swift
@Test func `The aquarium's dolphin nursery works`() async {
    let subject = Aquarium()
    Task {
        await subject.raiseDolphins()
    }
    await confirmation(until: .firstPass) {
        await subject.dolphins.count == 1
    }
}
```

> With the definition of `Aquarium` above, the closure will only need to be

Reviewer: nit: "may" only need? Since technically you can't predict or guarantee any particular timing; the system could be under heavy load
```suggestion
With the definition of `Aquarium` above, the closure may only need to be
```

evaluated a few times before it starts returning true. At which point polling
will end, and no failure will be reported.

> Polling will be stopped in the following cases:
> 
> - The specified `duration` has elapsed.
> - If the task that started the polling is cancelled.
> - For `PollingStopCondition.firstPass`: The first time the closure returns true
>   or a non-nil value
> - For `PollingStopCondition.stopsPassing`: The first time the closure returns
>   false or nil.
> - The first time the closure throws an error.

Reviewer: It might not be clear to a reader that this is an "OR" list not "AND". I find it helpful when proposals clarify that kind of thing by adding _"…, or"_ to the second-to-last bullet. For example:

```suggestion
Polling will be stopped when either:

- the specified duration has elapsed,
- the task that started the polling is cancelled,
- the closure returns a value that satisfies the stopping condition, or
- the closure throws an error.
```

(And in that suggestion I shortened the third bullet, instead of reiterating the stopping condition behaviors, since those are specified elsewhere in the document.)

///     with a matching stopCondition.
///     If no such trait has been added, then polling will wait at least
///     1 millisecond between polling attempts.
///     `interval` must be greater than 0.
///   - isolation: The actor to which `body` is isolated, if any.
///   - sourceLocation: The location in source where the confirmation was called.
///   - body: The function to invoke. The expression is considered to pass if
///     the `body` returns true. Similarly, the expression is considered to fail
///     if `body` returns false.
///
/// - Throws: A `PollingFailedError` if the `body` does not return true within
///   the polling duration.
///
/// Use polling confirmations to check that an event while a test is running in
/// complex scenarios where other forms of confirmation are insufficient. For
/// example, waiting on some state to change that cannot be easily confirmed
/// through other forms of `confirmation`.
@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
public func confirmation(
  _ comment: Comment? = nil,
  until stopCondition: PollingStopCondition,
  within duration: Duration? = nil,
  pollingEvery interval: Duration? = nil,

>   isolation: isolated (any Actor)? = #isolation,

Reviewer 2: Prefer expressing this by making the function “nonisolated(nonsending)”

Author: Thanks! I clearly missed the SE proposal that changed this.

  sourceLocation: SourceLocation = #_sourceLocation,

>   _ body: @escaping () async throws -> Bool

Reviewer 2: Should this be nonisolated nonsending as well?
